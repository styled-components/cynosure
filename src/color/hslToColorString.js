// @flow

import hsl from './hsl'
import hsla from './hsla'
import type { HslColor, HslaColor } from '../types/color'

/**
 * Converts a HslColor or HslaColor object to a color string.
 * This util is useful in case you only know on runtime which color object is
 * used. Otherwise we recommend to rely on `hsl` or `hsla`.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *   background: hslToColorString({ hue: 240, saturation: 1, lightness: 0.5 }),
 *   background: hslToColorString({ hue: 360, saturation: 0.75, lightness: 0.4, alpha: 0.72 }),
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   background: ${hslToColorString({ hue: 240, saturation: 1, lightness: 0.5 })};
 *   background: ${hslToColorString({ hue: 360, saturation: 0.75, lightness: 0.4, alpha: 0.72 })};
 * `
 *
 * // CSS in JS Output
 * element {
 *   background: "#00f";
 *   background: "rgba(179,25,25,0.72)";
 * }
 */
function hslToColorString(color: HslColor | HslaColor | number): string {
  if (
    typeof color === 'object' &&
    typeof color.hue === 'number' &&
    typeof color.saturation === 'number' &&
    typeof color.lightness === 'number'
  ) {
    // $FlowIgnoreNextLine not sure why this complains
    return (color.alpha && typeof color.alpha === 'number') ? hsla(color) : hsl(color)
  }

  throw new Error('Passed invalid argument to hslToColorString, please pass a HslColor or HslaColor object.')
}

export default hslToColorString
