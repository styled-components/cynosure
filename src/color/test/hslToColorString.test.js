// @flow
import hslToColorString from '../hslToColorString'

describe('hslToColorString', () => {
  it('should convert a HslColor to a reduced hex string', () => {
    expect({
      background: hslToColorString({ hue: 240, saturation: 1, lightness: 0.5 }),
    }).toMatchSnapshot()
  })

  it('should convert a HslColor to a hex string', () => {
    expect({
      background: hslToColorString({ hue: 360, saturation: 0.75, lightness: 0.4 }),
    }).toMatchSnapshot()
  })

  it('should convert a HslaColor to a rgba string', () => {
    expect({
      background: hslToColorString({
        hue: 360,
        saturation: 0.75,
        lightness: 0.4,
        alpha: 0.72,
      }),
    }).toMatchSnapshot()
  })

  it('should throw an error if anything else than a HslColor or HslaColor is provided', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    expect(() => hslToColorString({ red: 255, green: 1, hue: 240 })).toThrow(
      'Passed invalid argument to hslToColorString, please pass a HslColor or HslaColor object.',
    )
  })
})
