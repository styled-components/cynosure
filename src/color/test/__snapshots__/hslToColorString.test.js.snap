exports[`hslToColorString should convert a HslColor to a hex string 1`] = `
Object {
  "background": "#b31919",
}
`;

exports[`hslToColorString should convert a HslColor to a reduced hex string 1`] = `
Object {
  "background": "#00f",
}
`;

exports[`hslToColorString should convert a HslaColor to a rgba string 1`] = `
Object {
  "background": "rgba(179,25,25,0.72)",
}
`;
